### PR TITLE
Memory commit threshold in MB at which to create a dump.

### PR DESCRIPTION
**Pull request by kishanAnem**
Hi @lowleveldesign and @goldshtn 

will you  suggest some better option than 'p' in CommandLineOptions.cs line 31

`[Option('p', HelpText = "Memory commit threshold in MB at which to create a dump.")]`
`public int? MemoryCommitThreshold { get; set; }`

will you please guide me better approach than what i Implemented.  in MiniDumper.cs 

`public void MemoryCommitThreshold(int commitThreshold)`
 `private void OnTimedEvent(Object source, ElapsedEventArgs e, int commitThreshold)`

Thanks
kishan

### CODE REVIEW

**Comment by lowleveldesign** [comment]
Hi @kishan3034,

Thanks for the pull request. I'm afraid though that the feature to make a dump on a committed memory threshold is harder to implement. There is [an excellent post](http://blogs.microsoft.co.il/sasha/2016/01/05/Windows-process-memory-usage-demystified/) by @goldshtn explaining the Windows memory model, in which you may read that committed memory is more than only private bytes. I guess that to correctly measure it, we would need to enumerate memory pages reserved by the process and list those which are committed (maybe using [VirtualQueryEx](https://msdn.microsoft.com/en-us/library/windows/desktop/aa366907(v=vs.85).aspx)). But I would like to make some research on how it is implemented in procdump. Or maybe Sasha could give a hint here. I'm not writing all this to discourage you, as I also need to learn how this should be implemented. I propose to suspend this pull request for a while till we do the necessary research.

Regarding the command line options I would like to stick to the procdump command line, so people used to it could easily switch to minidumper. The control options for memory would be then:

```
-m      Memory commit threshold in MB at which to create a dump.
-ml     Trigger when memory commit drops below specified MB value.
```

But... if you feel you have some fuel for coding I would propose implementing the performance counter monitoring:

```
-p      Trigger on the specified performance counter when the threshold
          is exceeded. Note: to specify a process counter when there are
          multiple instances of the process running, use the process ID
          with the following syntax: "\Process(<name>_<pid>)\counter"
-pl     Trigger when performance counter falls below the specified value.
```

Whole API for querying performance counters is available under the System.Diagnostics namespace (no need for interoping). The tricky part might be to correctly assign name and pid to a counter instance (the API uses numbers to identify a counter instance and you need to query a special `ID Process` counter to match those ids). You may have a look at the code I implemented some time ago in the Musketeer monitoring service: https://github.com/lowleveldesign/diagnostics-kit/blob/master/LowLevelDesign.Diagnostics.Musketeer/Jobs/ServiceMonitorJob.cs or check [the post](https://lowleveldesign.org/2013/06/21/find-performance-counter-instance-by-process-id/). I would say the performance counters feature is more important, as the Process(*) counters set contains counters to monitor process memory. What do you think @kishan3034, @goldshtn?

**Comment by kishanAnem** [comment]
hi @lowleveldesign  and  @sasha,
Thanks for the review and guiding. definitely i'll work on performance counters.

I'm looking for a mentor, felt you're the right persons (@lowleveldesign and @goldshtn), will  you please be my mentor.

but I've doubt in [Defrag Tools: #10 - ProcDump - Triggers](https://channel9.msdn.com/shows/defrag-tools/defrag-tools-10-procdump-triggers)  at 16:53 Andrew Richards explained trigger -m, he took example notepad. to find private bytes used procexp. please check that based on that I proposed this pull request. please correct me if I'm wrong.

**Comment by lowleveldesign** [comment]
Hi @kishan3034,

You are right that from the video it does look like they are checking the private memory, but I also notice that `dumpbin /imports procdump` lists `VirtualQueryEx` as one of the imported functions. I guess that they somewhere use it. In the end we will need to reverse it a bit. 

I will contact you directly on twitter regarding the mentoring question.

**Comment by kishanAnem** [comment]
Hi @lowleveldesign and @sasha,

Is this PR fine?
Will you please let me know where we need to reverse it a bit?

File: MiniDumper/CommandLineOptions.cs [hunk 1]
```diff
@@ -28,6 +28,9 @@ class CommandLineOptions
         [Option('l', HelpText = "Display the debug logging of the process + diagnostics info from the minidumper.")]
         public bool Verbose { get; set; }
 
+        [Option('p', HelpText = "Memory commit threshold in MB at which to create a dump.")]
```
Comment (lowleveldesign) [inline MiniDumper/CommandLineOptions.cs]: I would prefer **-m** option (or **-mx** as we might have problems with a single-letter option in the CommandLineParser). If possible, please also add the **-ml** option to make dump when the memory usage gets below the specified threshold.

File: MiniDumper/CommandLineOptions.cs [hunk 2]
```diff
@@ -28,6 +28,9 @@ class CommandLineOptions
...
+        public int? MemoryCommitThreshold { get; set; }
```
Comment (lowleveldesign) [inline MiniDumper/CommandLineOptions.cs]: Could we rename it to `MemoryCommitThresholdInMB` and change type to `uint?`? Everyone will then know what is the unit.

File: MiniDumper/MiniDumper.cs [hunk 3]
```diff
@@ -178,6 +180,28 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
             }
         }
 
+        public void MemoryCommitThreshold(int commitThreshold)
+        {
+            System.Timers.Timer aTimer = new System.Timers.Timer();
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: This timer might stop working at any time - it is eligible for disposal when this method finishes. Please make it a field of the class and move the initialization to the constructor. We should also have a nice name for it (eg. `pollingTimer`) as we will probably also use it to query performance counters.

File: MiniDumper/MiniDumper.cs [hunk 4]
```diff
@@ -178,6 +180,28 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
...
+            aTimer.Interval = 1000;
+            aTimer.Elapsed += (sender, e) => OnTimedEvent(sender, e, commitThreshold); ;
+            aTimer.AutoReset = true;
+            aTimer.Enabled = true;
+        }
+
+        private void OnTimedEvent(Object source, ElapsedEventArgs e, int commitThreshold)
+        {
+           var process = Process.GetProcessById(pid);
+            double privateMemory = (process.PrivateMemorySize64 / 1024) / 1024;
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: I would prefer to convert the command line parameter to bytes (`MemoryCommitThresholdInMB * 1024 * 1024` and assign to a `UInt64` variable) and later compare it with the process memory usage - this way we are not loosing the last bytes of the gathered value. 

Also please use the `GetProcessMemoryInfo` method to query the process memory:

```
[StructLayout(LayoutKind.Sequential)]
private struct PROCESS_MEMORY_COUNTERS_EX
{
	public uint cb;
	public uint PageFaultCount;
	public UIntPtr PeakWorkingSetSize;
	public UIntPtr WorkingSetSize;
	public UIntPtr QuotaPeakPagedPoolUsage;
	public UIntPtr QuotaPagedPoolUsage;
	public UIntPtr QuotaPeakNonPagedPoolUsage;
	public UIntPtr QuotaNonPagedPoolUsage;
	public UIntPtr PagefileUsage;
	public UIntPtr PeakPagefileUsage;
	public UIntPtr PrivateUsage;
}

[DllImport("psapi.dll", SetLastError = true)]
static extern bool GetProcessMemoryInfo(IntPtr hProcess, out PROCESS_MEMORY_COUNTERS_EX counters, uint size);

EXAMPLE:

	PROCESS_MEMORY_COUNTERS_EX counters;
	counters.cb = (uint)Marshal.SizeOf<PROCESS_MEMORY_COUNTERS_EX>();
	var result = GetProcessMemoryInfo(processHandle, out counters, counters.cb);
	if (!result) {
		// TODO throw exception with last error
	} 
``` 

The `PrivateUsage` field in the structure is the one that interests us - use `ToUInt64` method to convert it to a number.

File: MiniDumper/MiniDumper.cs [hunk 5]
```diff
@@ -178,6 +180,28 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
...
+            if (privateMemory >= commitThreshold)
+            {
+                var timer = source as System.Timers.Timer;
+                timer.Stop();
+                timer.Dispose();
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: We should not dispose this timer as we will move it to the class field. And here we should also take into consideration the number of dumps the user wants (**-n** parameter).

File: MiniDumper/MiniDumper.cs [hunk 6]
```diff
@@ -178,6 +180,28 @@ public void PrintDebugString(OUTPUT_DEBUG_STRING_INFO outputDbgStrInfo)
             }
         }
 
+        public void MemoryCommitThreshold(int commitThreshold)
+        {
+            System.Timers.Timer aTimer = new System.Timers.Timer();
+            aTimer.Interval = 1000;
```
Comment (lowleveldesign) [inline MiniDumper/MiniDumper.cs]: I would like to have the interval configurable (as in procdump) using the **-s** option.

**Review by lowleveldesign — CHANGES_REQUESTED** [review]
I made few request changes - please have a look at them.

I also emailed Andrew Richards and he responded that procdump is using `GetProcessMemoryInfo` method to query process memory - let's use the same method (I added the pinvoke data in the review comment).

File: MiniDumper/MiniDumper.cs (hunk 6 above)
Comment (kishanAnem) [reply MiniDumper/MiniDumper.cs]: this interval for pooling interval. in procdump also they used 1000 ms. they said in Defrag tools.

-s | Consecutive seconds before dump is written (default is 10)
i think this is different.

Comment (lowleveldesign) [reply MiniDumper/MiniDumper.cs]: Ach yes, you are right. This interval just configures the minimum amount of time between subsequent dumps (when of course more than one dumps was requested). So we do not dump on every second.

**Pull request closed without merging**